CMakeLists.txt

```diff
@@ -0,0 +1,114 @@
+    # instead of system default for compatibility.
+    find_package(Matlab REQUIRED)
+    find_library(OpenMP_iomp5_LIBRARY NAMES iomp5 HINTS ${Matlab_ROOT_DIR}/sys/os/ PATH_SUFFIXES glnxa64 maci64)
+    find_library(OpenMP_pthreads_LIBRARY NAMES pthread CMAKE_FIND_ROOT_PATH_BOTH)
+endif()
+
+# Find required packages
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
+find_package(FFTW REQUIRED)
+
+if(FINUFFT_USE_OPENMP)
+    find_package(OpenMP REQUIRED)
+endif()
+
+set(FINUFFT_FFTW_LIBRARIES "FFTW::Float" "FFTW::Double" "FFTW::Float${FINUFFT_FFTW_SUFFIX}" "FFTW::Double${FINUFFT_FFTW_SUFFIX}")
+
+# This set of sources is compiled twice, once in single precision and once in double precision
+# The single precision compilation is done with -DSINGLE
+set(FINUFFT_PRECISION_DEPENDENT_SOURCES src/finufft.cpp src/simpleinterfaces.cpp src/spreadinterp.cpp src/utils.cpp fortran/finufftfort.cpp)
+
+# Set of compilers which behave like gcc
+set(FINUFFT_GNU_LIKE_COMPILERS AppleClang Clang GNU)
+
+# Utility function to enable ASAN on debug builds
+function(enable_asan target)
+    if(NOT FINUFFT_ENABLE_SANITIZERS)
+        return()
+    endif()
+
+    if(CMAKE_CXX_COMPILER_ID IN_LIST FINUFFT_GNU_LIKE_COMPILERS)
+        # Enable only on clang / gcc compilers.
+        target_compile_options(${target} PRIVATE $<$<CONFIG:DEBUG>:-fsanitize=address>)
+        target_link_options(${target} PRIVATE $<$<CONFIG:DEBUG>:-fsanitize=address>)
+    endif()
+endfunction()
+
+# Utility function to set finufft compilation options.
+function(set_finufft_options target)
+    set_property(TARGET ${target} PROPERTY POSITION_INDEPENDENT_CODE ON)
+    set_property(TARGET ${target} PROPERTY CMAKE_CXX_STANDARD 14)
+    enable_asan(${target})
+    target_compile_options(${target} PRIVATE SHELL:$<$<CONFIG:Release,RelWithDebInfo>:${FINUFFT_ARCH_FLAGS}>)
+    target_include_directories(${target} PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}/include")
+    if(FINUFFT_USE_OPENMP)
+        target_link_libraries(${target} PRIVATE OpenMP::OpenMP_CXX)
+    else()
+        if(CMAKE_CXX_COMPILER_ID IN_LIST FINUFFT_GNU_LIKE_COMPILERS)
+            # OpenMP disabled, suppress unknown pragma warnings to avoid spam.
+            target_compile_options(${target} PRIVATE "-Wno-unknown-pragmas")
+        endif()
+    endif()
+endfunction()
+
+# Main finufft libraries
+add_library(finufft_f32 OBJECT ${FINUFFT_PRECISION_DEPENDENT_SOURCES})
+target_compile_definitions(finufft_f32 PRIVATE SINGLE)
+set_finufft_options(finufft_f32)
+target_link_libraries(finufft_f32 PUBLIC ${FINUFFT_FFTW_LIBRARIES})
+
+add_library(finufft_f64 OBJECT ${FINUFFT_PRECISION_DEPENDENT_SOURCES})
+set_finufft_options(finufft_f64)
+target_link_libraries(finufft_f64 PUBLIC ${FINUFFT_FFTW_LIBRARIES})
+
+add_library(finufft SHARED src/utils_precindep.cpp contrib/legendre_rule_fast.c)
+set_finufft_options(finufft)
+target_link_libraries(finufft PUBLIC finufft_f32 finufft_f64 m)
```

Comment:
If I am not mistaken OBJECT libraries should not be used to link:
https://gitlab.kitware.com/cmake/community/-/wikis/doc/tutorials/Object-Library

Reply:
I think this documentation is very out of date. Here's the documentation I see for `target_link_libraries`: https://cmake.org/cmake/help/latest/command/target_link_libraries.html#linking-object-libraries

In particular, it says:
"Normal libraries and executables may link to [Object Libraries](https://cmake.org/cmake/help/latest/manual/cmake-buildsystem.7.html#object-libraries) to get their **objects** and **usage requirements**."

The `TARGET_OBJECTS` syntax is also fully supported since very recently, but I believe it is mostly intended to support scenarios to create a transitive dependency on the object files themselves, which is not needed here.

Reply:
Thanks for updating me. Then I agree with your implementation.